### DISCUSSION TITLE
embedded-graphics: How to properly use the framebuffer

### DISCUSSION THREAD

**Discussion by TheTechRobo** [rust-wii/ogc-rs#108] [Q&A] [answered]
Hello!

I'm trying to make a simple 2D demo right now before I do anything else with this, just to get familiar with it. I didn't want to touch GX or writing C bindings to e.g. grrlib, so I'm using the embedded-graphics demo as a starting point. I haven't been able to find much concrete info about video in libogc on Google. :-(

My question is: How do I properly handle the two aspect ratios, and NTSC vs PAL? From my experimentation, I believe NTSC (as expected) uses only the first 480 lines of the framebuffer. Positioning things past y=480 does not show up on the screen. Does this mean that PAL will simply extend the amount of lines used to 576? That wouldn't make sense, since the framebuffer seems to be limited to 528 lines. Finally, how would widescreen work if I only have 640 horizontal pixels?

Thank you in advance for any help!

**Comment by ProfElements** [comment]
> How do I properly handle the two aspect ratios, and NTSC vs PAL

You will want to change the render config handed to the `setup` func
Currently you will have to enable `ffi` to get anything besides the the preferred mode which should be able to handle it. 

> Does this mean that PAL will simply extend the amount of lines used to 576?

The 576 lines include under/overscan areas usually hidden by most crts so its uses 528 lines as  a roughly safe area. 

> Finally, how would widescreen work if I only have 640 horizontal pixels?
Widescreen right now is very thought out. But from what I remember widescreen stretches the pixels into rectangles isntead of squares. Though this might be wrong.

As a warning some things just currently don't work with the  embedded-graphics examples:

 
Not working with a circle: https://github.com/rust-wii/ogc-rs/issues/43

**Reply by TheTechRobo** [reply]
So on PAL, is it extended to 528 lines? How do I detect PAL vs NTSC?

**Reply by ProfElements** [reply, marked as answer]
On Pal it does extend to 528 lines.

To get the preferred format you would need to do something like this 
 ```rs
fn get_video_format() -> String {
    if let Ok(fd) = ios::open(c"/title/00000001/00000002/data/setting.txt", Mode::Read) {
        const GET_FILE_STATS: i32 = 11;
        let mut out_buf = [0u8; 8];
        let (size, _seek_pos) = if ios::ioctl(fd, GET_FILE_STATS, &[], &mut out_buf).is_ok() {
            (
                usize::try_from(u32::from_be_bytes(out_buf[0..4].try_into().unwrap())).unwrap(),
                usize::try_from(u32::from_be_bytes(out_buf[4..8].try_into().unwrap())).unwrap(),
            )
        } else {
            (0usize, 0usize)
        };

        let mut bytes = vec![0u8; size];

        if let Ok(bytes_read) = ios::read(fd, &mut bytes) {
            unsafe {
                bytes.set_len(bytes_read.try_into().unwrap());
            }
        }

        let mut key: u32 = 0x73B5DBFA;
        for byte in &mut *bytes {
            *byte ^= u8::try_from(key & 0xff).unwrap();
            key = (key << 1) | (key >> 31);
        }

        let text = if let Err(vld) = core::str::from_utf8(&bytes) {
            unsafe { core::str::from_utf8_unchecked(&bytes[..vld.valid_up_to()]) }
        } else {
            ""
        };

        for line in text.lines() {
            if let Some(char) = line.find("VIDEO=") {
                return line[char + 6..].trim().to_string();
            }
        }
        let _ = ios::close(fd);
    }

    return String::from("");
}
```
Where this return "NTSC", "PAL", or "MPAL"
